### PR TITLE
Add ability to output all keys and values in thread context

### PR DESCRIPTION
**Pull request by rhanneken**
### Description

The keys in a thread context can differ for different code paths. It would be good if one could log whatever keys and values exist in the thread context without specifying all possible keys in the logging configuration.

This change modifies the `context` placeholder so that it can be used without specifying a key. When no key is specified, all keys and their values are sent to the output.

Linked website pull request, if the documention of https://tinylog.org/v2/ has to be updated:

tinylog-org/website#188 

### Definition of Done

- [X] I read [contributing.md](https://github.com/tinylog-org/tinylog/blob/v2.7/contributing.md)
- [X] There are no TODOs left in the code
- [X] Code style follows the tinylog standard
- [X] All classes and methods have Javadoc
- [X] Changes are covered by JUnit tests including edge cases, errors, and exception handling
- [X] Maven build works including compiling, tests, and checks (`mvn verify`)
- [X] Changes are committed by a verified email address that is assigned to the GitHub account (https://github.com/settings/emails)
- [X] Documentation on the [tinylog website](https://tinylog.org/v2/) is up-to-date or updated by a pull request to the repository [tinylog-org/website](https://github.com/tinylog-org/website)

### Agreements

- [X] I agree that my changes will be published under the terms of the [Apache License 2.0](https://github.com/tinylog-org/tinylog/blob/v2.0/license.txt) (mandatory)
- [X] I agree that my GitHub user name will be published in the release notes (optional)

### CODE REVIEW

File: tinylog-impl/src/test/java/org/tinylog/pattern/FormatPatternParserTest.java [hunk 1]
```diff
@@ -169,12 +169,14 @@ public void context() {
 	}
 
 	/**
-	 * Verifies that {@code {context}} without a defined key will produce an error.
+	 * Verifies that {@code {context}} without a defined key will output all keys and their values.
 	 */
 	@Test
 	public void contextMissingKey() {
```
Comment (pmwmedia) [inline tinylog-impl/src/test/java/org/tinylog/pattern/FormatPatternParserTest.java]: Can you rename this test method?

File: tinylog-impl/src/main/java/org/tinylog/pattern/ThreadContextToken.java [hunk 2]
```diff
@@ -58,21 +68,39 @@ public Collection<LogEntryValue> getRequiredLogEntryValues() {
...
+			String value = logEntry.getContext().get(key);
+			if (value == null) {
+				builder.append(defaultValue);
+			} else {
+				builder.append(value);
+			}
 		} else {
-			builder.append(value);
+			boolean first = true;
+			for (Map.Entry<String, String> contextEntry : logEntry.getContext().entrySet()) {
```
Comment (pmwmedia) [inline tinylog-impl/src/main/java/org/tinylog/pattern/ThreadContextToken.java]: Can you sort the keys alphabetically?

File: tinylog-impl/src/main/java/org/tinylog/pattern/ThreadContextToken.java [hunk 3]
```diff
@@ -58,21 +68,39 @@ public Collection<LogEntryValue> getRequiredLogEntryValues() {
...
+			}
 		}
 	}
 
 	@Override
 	public void apply(final LogEntry logEntry, final PreparedStatement statement, final int index) throws SQLException {
-		String value = logEntry.getContext().get(key);
-		if (value == null && !DEFAULT_EMPTY_VALUE.equals(defaultValue)) {
-			statement.setString(index, defaultValue);
+		if (key == null || key.isEmpty()) {
```
Comment (pmwmedia) [inline tinylog-impl/src/main/java/org/tinylog/pattern/ThreadContextToken.java:95]: Can you add a JUnit test for this case? Currently line 93 - 96 are not covered by any test.

**Review by pmwmedia — CHANGES_REQUESTED** [review]
Thank you for your commit! It looks great so far :) I have just three minor improvement suggestions.

**Comment by rhanneken** [comment]
@pmwmedia Thanks for your feedback! I've made some new changes. Let me know if I got it right.

File: tinylog-impl/src/test/java/org/tinylog/pattern/FormatPatternParserTest.java (hunk 1 above)
Comment (rhanneken) [reply tinylog-impl/src/test/java/org/tinylog/pattern/FormatPatternParserTest.java]: I renamed it to `contextWithoutKey`.

File: tinylog-impl/src/main/java/org/tinylog/pattern/ThreadContextToken.java (hunk 2 above)
Comment (rhanneken) [reply tinylog-impl/src/main/java/org/tinylog/pattern/ThreadContextToken.java]: Done!

File: tinylog-impl/src/main/java/org/tinylog/pattern/ThreadContextToken.java (hunk 3 above)
Comment (rhanneken) [reply tinylog-impl/src/main/java/org/tinylog/pattern/ThreadContextToken.java:95]: Done!

**Comment by pmwmedia** [comment]
Thank you very much! I will publish it together with tinylog 2.8.

**Comment by rhanneken** [comment]
Great! Do you need me to make an equivalent change to the v3.0 branch?

**Comment by pmwmedia** [comment]
I would be very happy about a PR for the v3.0 branch :)
